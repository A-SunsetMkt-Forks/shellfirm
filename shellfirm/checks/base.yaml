- from: base
  test: ':\(\)\{ :|:& \};:'
  description: "This short line defines a shell function that creates new copies of itself.\nThe process continually replicates itself, and its copies continually replicate themselves, quickly taking up all your CPU time and memory.\nThis can cause your computer to freeze. Itâ€™s basically a denial-of-service attack."
  id: base:bash_fork_bomb
- from: base
  test: \s*crontab\s+-r
  description: "You are going to remove your entire table of cron tasks."
  id: base:delete_all_cron_tasks
- from: base
  test: \s*history(.*)[|](.*)(bash|sh)($|\s)
  description: "You are going to executes every command from the command log that you have already executed."
  id: base:execute_all_history_commands
- from: base
  test: reboot(\s|$)
  description: "You are going to reboot your machine."
  id: base:reboot_machine
- from: base
  test: shutdown(\s|$)
  description: "You are going to shutdown your machine."
  id: base:shutdown_machine
