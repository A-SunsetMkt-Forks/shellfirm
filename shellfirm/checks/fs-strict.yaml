- from: fs-strict
  test: chmod\s
  description: "Are you sure that you want to change permissions?"
  id: fs-strict:change_permissions
- from: fs-strict
  test: rm\s
  description: "Are you sure you want to continue with deletion?"
  id: fs-strict:any_deletion