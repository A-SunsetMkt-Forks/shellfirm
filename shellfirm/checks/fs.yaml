- from: fs
  test: 'rm\s{1,}(-r|-fr|-rf)\s*(\*|\.{2,}|/)'
  method: Regex
  description: "You are going to delete everything in the path."
  id: fs:recursively_delete
- from: fs
  test: mv\s{1,}([a-zA-Z0-9.!@#$%^&*()']*) /dev/null
  method: Regex
  description: "The files will be discarded and destroyed."
  id: fs:move_to_dev_null
- from: fs
  test: .*>(.*)
  method: Regex
  description: "The above command is used to flush the content of a file."
  filters:
    IsFileExists: "1"
  id: fs:flush_file_content
- from: fs
  test: chmod\s{1,}(-R|--recursive)\s{1}(\*|\.{2,}|/)
  method: Regex
  description: "Change permission to all root files can brake your some thinks like SSH keys."
  id: fs:recursively_chmod
- from: fs
  test: find\s.*-delete
  method: Regex
  description: "Did you -delete flag in the wrong order? find -delete going to delete all the file under your current path."
  id: fs:delete_find_files
