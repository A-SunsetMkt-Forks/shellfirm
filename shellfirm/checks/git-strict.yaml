- from: git-strict
  test: git\s{1,}add\s{1,}.
  description: "Do you want to add all changes files to the staging area?"
  id: git-strict:git_add_all
- from: git-strict
  test: git\s{1,}commit\s{1,}(--all|-(.*a))
  description: "Do you want to commit all changes file and commit them?"
  id: git-strict:git_commit_all