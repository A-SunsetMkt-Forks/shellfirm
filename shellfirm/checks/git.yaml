- from: git
  test: git reset
  method: Contains
  description: "This command going to reset all your local changes."
  id: git:git_reset