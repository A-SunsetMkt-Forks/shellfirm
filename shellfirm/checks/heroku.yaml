- from: heroku
  test: heroku\s*ps:restart
  description: "Restarting app dynos"
  id: heroku:restart_app_dynos
- from: heroku
  test: heroku\s*ps:stop
  description: "Stop app dynos"
  id: heroku:stop_app_dynos
- from: heroku
  test: heroku\s*ps:kill
  description: "Kill app dynos"
  id: heroku:kill_app_dynos
- from: heroku
  test: heroku\s*maintenance:on
  description: "Put the app into maintenance mode?"
  id: heroku:enable_maintenance_mode
- from: heroku
  test: heroku\s*members:remove
  description: "Removes a user from a team"
  id: heroku:remove_member
- from: heroku
  test: heroku\s*features:disable
  description: "Disables an app feature"
  id: heroku:disable_app_feature
- from: heroku
  test: heroku\s*container:rm
  description: "Remove the process type from your app"
  id: heroku:remove_app_container
- from: heroku
  test: heroku\s*config:unset
  description: "unset one or more config vars"
  id: heroku:unset_environment_variable
- from: heroku
  test: heroku\s*clients:destroy
  description: "Delete client by ID"
  id: heroku:destroy_client
- from: heroku
  test: heroku\s*clients:rotate
  description: "Rotate OAuth client secret"
  id: heroku:rotate_oauth_client
- from: heroku
  test: heroku\s*clients:update
  description: "Update OAuth client"
  id: heroku:update_oauth_client
- from: heroku
  test: heroku\s*apps:destroy
  description: "Permanently destroy an app"
  id: heroku:destroy_app
- from: heroku
  test: heroku\s*apps:leave
  description: "Remove yourself from a team app"
  id: heroku:remove_yourself_from_app
- from: heroku
  test: heroku\s*apps:rename
  description: "Rename an app"
  id: heroku:rename_app_name
- from: heroku
  test: heroku\s*addons:destroy
  description: "Permanently destroy an add-on resource"
  id: heroku:destroy_addons
- from: heroku
  test: heroku\s*addons:detach
  description: "Detach an existing add-on resource from an app"
  id: heroku:detach_addon
- from: heroku
  test: heroku\s*access:remove
  description: "Remove users from a team app"
  id: heroku:remove_user_access
- from: heroku
  test: heroku\s*access:update
  description: "Update existing collaborators on an team app"
  id: heroku:update_collaborators_access
- from: heroku
  test: heroku\s*repo:reset
  description: "Reset heroku repo"
  id: heroku:reset_repo