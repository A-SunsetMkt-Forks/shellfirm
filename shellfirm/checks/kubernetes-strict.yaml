- from: kubernetes-strict
  test: \s*(k|kubectl)\s*(.*)delete[[:space:]][^ns]
  method: Regex
  enable: true
  description: "This command will going to delete a given resource."
- from: kubernetes-strict
  test: \s*(kubectl|k)\s*set
  method: Regex
  enable: true
  description: "This command will update the given given resource."
- from: kubernetes-strict
  test: \s*(kubectl|k)\s*scale
  method: Regex
  enable: true
  description: "This command will set a new size for a given resource."
- from: kubernetes-strict
  test: \s*(kubectl|k)\s*rollout
  method: Regex
  enable: true
  description: "This command will manage a rollout for a given resource."