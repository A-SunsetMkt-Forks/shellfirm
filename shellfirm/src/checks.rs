//! Manage command checks

///
use crate::config::{Challenge, Method};
use crate::prompt;
use anyhow::Result;
use console::style;
use log::debug;
use rayon::prelude::*;
use regex::Regex;
use serde_derive::{Deserialize, Serialize};
use std::collections::HashMap;

// list of custom filter
#[derive(Debug, Deserialize, Serialize, PartialEq, Eq, Hash, Clone)]
pub enum FilterType {
    IsFileExists,
}

/// Describe single check
#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct Check {
    /// test ia a value that we check the command.
    pub test: String,
    /// The type of the check.
    pub method: Method,
    /// boolean for ignore check
    pub enable: bool,
    /// description of what is risky in this command
    pub description: String,
    /// the group of the check see files in `checks` folder
    pub from: String,
    #[serde(default)]
    pub challenge: Challenge,
    #[serde(default)]
    pub filters: HashMap<FilterType, String>,
}

/// prompt a challenge to the user
///
/// # Errors
///
/// Will return `Err` when could not convert checks to yaml
pub fn challenge(challenge: &Challenge, checks: &[Check], dryrun: bool) -> Result<bool> {
    if dryrun {
        eprintln!("{}", serde_yaml::to_string(checks)?);
        return Ok(true);
    }
    eprintln!("{}", style("#######################").yellow().bold());
    eprintln!("{}", style("# RISKY COMMAND FOUND #").yellow().bold());
    eprintln!("{}", style("#######################").yellow().bold());

    let mut descriptions: Vec<String> = Vec::new();
    for check in checks {
        if !descriptions.contains(&check.description) {
            descriptions.push(check.description.to_string());
        }
    }
    for description in descriptions {
        eprintln!("* {}", description);
    }
    eprintln!();

    let show_challenge = challenge;

    Ok(match show_challenge {
        Challenge::Math | Challenge::Default => prompt::math_challenge(),
        Challenge::Enter => prompt::enter_challenge(),
        Challenge::Yes => prompt::yes_challenge(),
    })
}

/// Check if the given command matched to on of the checks
///
/// # Arguments
///
/// * `checks` - List of checks that we want to validate.
/// * `command` - Command check.
#[must_use]
pub fn run_check_on_command(checks: &[Check], command: &str) -> Vec<Check> {
    checks
        .par_iter()
        .filter(|&v| v.enable)
        .filter(|&v| is_match(v, command))
        .filter(|&v| check_custom_filter(v, command))
        .map(std::clone::Clone::clone)
        .collect()
}

/// filter custom checks
///
/// When true is returned it mean the filter should keep the check and not filter our the check.
///
/// # Arguments
///
/// * `check` - Check struct
/// * `command` - Command.
fn check_custom_filter(check: &Check, command: &str) -> bool {
    // if check is not regex type or custom filters are not configure return true.
    if check.method != Method::Regex || check.filters.is_empty() {
        return true;
    }
    // Capture command groups from the current check
    let caps = Regex::new(&check.test).unwrap().captures(command).unwrap();

    // by default true is return. it mean the check not filter out (safe side security).
    let mut keep_check = true;
    for (filter_type, filter_params) in &check.filters {
        debug!(
            "filter information: command {} include filter: {:?} filter_params: {}",
            command, filter_type, filter_params
        );

        let keep_filter = match filter_type {
            FilterType::IsFileExists => filter_is_file_exists(
                caps.get(filter_params.parse().unwrap())
                    .map_or("", |m| m.as_str()),
            ),
        };

        if !keep_filter {
            keep_check = false;
            break;
        }
    }

    keep_check
}

/// Check if the given command match to one of the existing checks
///
/// # Arguments
///
/// * `check` - Check struct
/// * `command` - command for the check
fn is_match(check: &Check, command: &str) -> bool {
    match check.method {
        Method::Contains => is_contains(&check.test, command),
        Method::StartWith => is_start_with(&check.test, command),
        Method::Regex => is_regex(&check.test, command),
    }
}

/// Is the command contains the given check.
///
/// # Arguments
///
/// * `test` - check value
/// * `command` - command value
fn is_contains(test: &str, command: &str) -> bool {
    command.contains(test)
}

/// is the command start with the given check.
///
/// # Arguments
///
/// * `test` - check value
/// * `command` - command value
fn is_start_with(test: &str, command: &str) -> bool {
    command.starts_with(test)
}

/// Is the command match to the given regex.
///
/// # Arguments
///
/// * `test_r` - check value
/// * `command` - command value
fn is_regex(test_r: &str, command: &str) -> bool {
    Regex::new(test_r).unwrap().is_match(command)
}

/// check if the path exists (file and folder).
///
/// # Arguments
///
/// * `file_path` - check path.
fn filter_is_file_exists(file_path: &str) -> bool {
    let mut file_path: String = file_path.trim().into();
    if file_path.starts_with('~') {
        match dirs::home_dir() {
            Some(path) => {
                file_path = file_path.replace('~', &path.display().to_string());
            }
            None => return true,
        };
    }
    debug!("check is file {} exists", file_path);
    return std::path::Path::new(file_path.trim()).exists();
}

#[cfg(test)]
mod test_checks {
    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn is_match_command() {
        let regex_check = Check {
            test: String::from("rm.+(-r|-f|-rf|-fr)*"),
            method: Method::Regex,
            enable: true,
            description: String::from(""),
            from: String::from(""),
            challenge: Challenge::Default,
            filters: HashMap::new(),
        };
        let contains_check = Check {
            test: String::from("test"),
            method: Method::Contains,
            enable: true,
            description: String::from(""),
            from: String::from(""),
            challenge: Challenge::Default,
            filters: HashMap::new(),
        };
        let startwith_check = Check {
            test: String::from("start"),
            method: Method::StartWith,
            enable: true,
            description: String::from(""),
            from: String::from(""),
            challenge: Challenge::Default,
            filters: HashMap::new(),
        };
        assert_debug_snapshot!(&regex_check);
        assert_debug_snapshot!(&contains_check);
        assert_debug_snapshot!(&startwith_check);
    }
    #[test]
    fn can_check_is_contains() {
        assert_debug_snapshot!(is_contains("test", "test is valid"));
        assert_debug_snapshot!(is_contains("test is valid", "not-found"));
    }

    #[test]
    fn can_check_is_start_with() {
        assert_debug_snapshot!(is_start_with("test is", "test is valid"));
        assert_debug_snapshot!(is_start_with("test is valid", "is"));
    }

    #[test]
    fn can_check_is_regex_match() {
        assert_debug_snapshot!(is_regex("rm.+(-r|-f|-rf|-fr)*", "rm -rf"));
        assert_debug_snapshot!(is_regex("^f", "rm -rf"));
    }
}
