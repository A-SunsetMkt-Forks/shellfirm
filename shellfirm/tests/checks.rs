use std::{fs, path::PathBuf};

use insta::assert_debug_snapshot;
use itertools::Itertools;
use serde_derive::Deserialize;
use shellfirm::checks::run_check_on_command;

#[derive(Debug, Deserialize, Clone)]
struct TestSensitivePatterns {
    pub test: String,
    pub description: String,
}

#[derive(Debug, Deserialize, Clone)]
struct TestSensitivePatternsResult {
    pub file_path: String,
    pub test: String,
    pub check_detection_ids: Vec<String>,
    pub test_description: String,
}

#[test]
fn test_checks() {
    let checks = shellfirm::checks::get_all().unwrap();

    let test_files_path = fs::read_dir("./tests/checks")
        .unwrap()
        .filter_map(|entry| entry.ok().map(|e| e.path()))
        .collect::<Vec<PathBuf>>();

    for file in test_files_path {
        let file_name = file.file_name().unwrap().to_str().unwrap().to_string();
        let mut test_file_results: Vec<TestSensitivePatternsResult> = Vec::new();
        let tests: Vec<TestSensitivePatterns> =
            serde_yaml::from_reader(std::fs::File::open(&file.display().to_string()).unwrap())
                .unwrap();

        for test in tests {
            let run_result = run_check_on_command(&checks, &test.test);

            test_file_results.push(TestSensitivePatternsResult {
                file_path: file_name.clone(),
                test: test.test,
                check_detection_ids: run_result
                    .iter()
                    .map(|f| f.id.to_string())
                    .sorted_by(|a, b| Ord::cmp(&b, &a))
                    .collect::<Vec<_>>(),
                test_description: test.description,
            });
        }
        assert_debug_snapshot!(file_name, test_file_results);
    }
}

#[test]
fn test_missing_patterns_coverage() {
    let checks = shellfirm::checks::get_all().unwrap();

    let test_files_path = fs::read_dir("./tests/checks")
        .unwrap()
        .filter_map(|entry| {
            entry
                .ok()
                .map(|e| e.file_name().to_str().unwrap().to_string())
        })
        .collect::<Vec<String>>();

    let mut not_covered = vec![];
    for check in checks {
        if !test_files_path.contains(&format!("{}.yaml", &check.id.replace(':', "-"))) {
            not_covered.push(check.id);
        }
    }
    assert_debug_snapshot!(not_covered);
}
