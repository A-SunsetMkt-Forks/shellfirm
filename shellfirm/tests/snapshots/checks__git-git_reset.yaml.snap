---
source: shellfirm/tests/checks.rs
expression: test_file_results
---
[
    TestSensitivePatternsResult {
        file_path: "git-git_reset.yaml",
        test: "git reset",
        check_detection_ids: [
            "git:git_reset",
        ],
        test_description: "match command",
    },
    TestSensitivePatternsResult {
        file_path: "git-git_reset.yaml",
        test: "git    reset",
        check_detection_ids: [
            "git:git_reset",
        ],
        test_description: "match command",
    },
    TestSensitivePatternsResult {
        file_path: "git-git_reset.yaml",
        test: "gitt reset",
        check_detection_ids: [],
        test_description: "invalid command",
    },
]
